server/src/main/java/io/deephaven/server/arrow/ArrowFlightUtil.java

```diff
@@ -66,6 +71,166 @@ public class ArrowFlightUtil {
+    }
+
+    private static void createAndSendSnapshot(BaseTable table, BitSet columns, RowSet viewport, boolean reverseViewport,
+            BarrageSnapshotOptions snapshotRequestOptions, StreamObserver<BarrageStreamGenerator.View> listener,
+            BarragePerformanceLog.SnapshotMetricsHelper metrics) {
+        // if the table is static or append-only and a full snapshot is requested, we can make and send multiple
+        // snapshots to save memory and operate more efficiently
+
+        if (viewport == null && (!table.isRefreshing() || table.isAddOnly())) {
+            // start with small value and grow
+            long snapshotTargetCellCount = MIN_SNAPSHOT_CELL_COUNT;
+            double snapshotNanosPerCell = 0.0;
+            long startKey = 0L;
+
+            final long columnCount =
+                    Math.max(1, columns != null ? columns.cardinality() : table.getDefinition().getColumns().length);
+
+            boolean isComplete = false;
+
+            try (final WritableRowSet snapshotViewport = RowSetFactory.empty()) {
+                while (!isComplete) {
+                    // compute the next range to snapshot
+                    final long cellCount =
+                            Math.max(MIN_SNAPSHOT_CELL_COUNT,
+                                    Math.min(snapshotTargetCellCount, MAX_SNAPSHOT_CELL_COUNT));
+                    long endKey = startKey + (cellCount / columnCount) - 1;
+
+                    // if we are approaching the end of the table, request all remaining rows
+                    if (endKey > table.size()) {
+                        endKey = Long.MAX_VALUE - 1;
+                        isComplete = true;
+                    }
+
+                    try (final RowSet snapshotPartialViewport =
+                            RowSetFactory.fromRange(startKey, endKey)) {
+                        snapshotViewport.insert(snapshotPartialViewport);
+
+                        // grab the snapshot and measure elapsed time for next projections
+                        long start = System.nanoTime();
+                        final BarrageMessage msg =
+                                ConstructSnapshot.constructBackplaneSnapshotInPositionSpace(log, table,
+                                        columns, snapshotPartialViewport, null);
+                        msg.modColumnData = ZERO_MOD_COLUMNS; // no mod column data for DoGet
+                        long elapsed = System.nanoTime() - start;
+                        // accumulate snapshot time in the metrics
+                        metrics.snapshotNanos += elapsed;
+
+                        // send out the data. Note that although a `BarrageUpdateMetaData` object will
+                        // be provided with each unique snapshot, vanilla Flight clients will ignore
+                        // these and see only an incoming stream of batches
+                        try (final BarrageStreamGenerator bsg = new BarrageStreamGenerator(msg, metrics)) {
```

Comment:
Can any SizeExceptions be thrown here or from `getSnapshotView`? Should we attempt to further break this request up if things get too large?